pkg/util/grpcclient/signing_handler.go

```diff
@@ -0,0 +1,95 @@
+package grpcclient
+
+import (
+	"context"
+
+	"github.com/weaveworks/common/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	reqSignHeaderName = "x-req-signature"
+)
+
+const (
+	ErrDifferentSignaturePresent = errors.Error("different signature already present")
+	ErrSignatureNotPresent       = errors.Error("signature not present")
+	ErrSignatureMismatch         = errors.Error("signature mismatch")
+)
+
+// SignRequest define the interface that must be implemented by the request structs to be signed
+type SignRequest interface {
+	// Sign returns the signature for the given request
+	Sign(context.Context) (string, error)
+	// VerifySign returns true if the signature is valid
+	VerifySign(context.Context, string) (bool, error)
+}
+
+func UnarySigningServerInterceptor(ctx context.Context, req interface{}, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+	rs, ok := req.(SignRequest)
+	if !ok {
+		return handler(ctx, req)
+	}
+
+	md, ok := metadata.FromIncomingContext(ctx)
+
+	if !ok {
+		return nil, ErrSignatureNotPresent
+	}
+
+	sig, ok := md[reqSignHeaderName]
+
+	if !ok || len(sig) != 1 {
+		return nil, ErrSignatureNotPresent
+	}
+
+	valid, err := rs.VerifySign(ctx, sig[0])
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !valid {
+		return nil, ErrSignatureMismatch
+	}
+
+	return handler(ctx, req)
+}
+
+func UnarySigningClientInterceptor(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	rs, ok := req.(SignRequest)
+
+	if !ok {
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+
+	signature, err := rs.Sign(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	md, ok := metadata.FromOutgoingContext(ctx)
+	if !ok {
+		md = metadata.New(map[string]string{})
+	}
+
+	newCtx := ctx
+
+	if s, ok := md[reqSignHeaderName]; ok {
+		if len(s) == 1 {
+			if s[0] != signature {
```

Comment:
nit: merged this `if` with the above if statement will save you on indentation, which makes the code look nicer :) 
like `if len(s) == 1 && s[0] != signature`  

or maybe just one gigantic if ... 

```
if len(s) != 1 || s[0] != signagure {
   return ErrDifferentSignaturePresent
}
```

Reply:
Then i need to check again on the else (len > 1), right?

Reply:
Oh i'm blind I thought you are returning `ErrDifferentSignaturePresent` in the case of len > 1 too. 

My bad.